### PR TITLE
add disposable launcher framework

### LINKED ISSUE #1103: Feature request: Disposable launchers framework
ACE3 and some other mods (incl. RHS) have scripted solution for disposable launchers like NLAW, AT-4, RPG-18. Sometimes running ACE3 is too much for semi-realism games, but disposable AT are needed.

Would be great to have some framework in CBA, that allow to plug my mod's custom launcher classes with minimal set up and to make them disposable.

Expected Minimum: To have all needed settings, scripts and handlers in CBA + some documentation ('how to' and what assets my mod should have). It would be super great to make behaviour switchable - so i can switch back to default vanilla behaviour while playing vanilla missions, where launchers are expected to be reloadable.

### PR DESCRIPTION
**Pull request by commy2**
**When merged this pull request will:**
- ports the BWA3 and ACE disposable launcher frameworks to CBA
- close #1103

- [x] doc
- [x] example configs
- [x] pick up loadable
- [x] replace with used tube
- [x] Arsenal

### CODE REVIEW

**Comment by commy2** [comment]
quick link for me: https://github.com/acemod/ACE3/tree/master/addons/disposable

**Comment by PabstMirror** [comment]
If we are starting over, I'd think we should require the fired event handler to be on the weapon
```
class nlaw 
class EventHandlers {
            fired = "call CBA_fnc_handleDisposable"
```

**Comment by commy2** [comment]
@PabstMirror You can't see it, but that is what I am doing in BWA3. Here is the WIP "doc" which is the state of the parallel branch on the BWA3 repository (where I test all this):

<details><summary>Click to expand</summary>

## Disposable Launchers

### Config examples

```cpp
#define LAUNCHER_MASS 80
#define LAUNCHER_MAGAZINE_MASS 40

class CBA_LoadedLaunchers {
    BWA3_PzF3[] = {"BWA3_PzF3_Loaded","BWA3_PzF3_Used"};
};

class CfgMagazines {
    class CA_LauncherMagazine;
    class BWA3_PzF3_Tandem: CA_LauncherMagazine {
        author = "$STR_BWA3_Author";
        scope = 1;

        mass = LAUNCHER_MAGAZINE_MASS;
        allowedSlots[] = {};
    };
};

class CfgWeapons {
    class Launcher;
    class Launcher_Base_F: Launcher {
        class WeaponSlotsInfo;
    };

    class BWA3_PzF3: Launcher_Base_F {
        author = "$STR_BWA3_Author";
        scope = 1;
        scopeArsenal = 1;
        baseWeapon = "BWA3_PzF3";

        magazines[] = {"BWA3_PzF3_Tandem"};
        magazineReloadTime = 0.1;

        class EventHandlers {
            fired = "_this call CBA_fnc_weaponEvents"; // this weapon eventhandler is required!
        };

        class WeaponSlotsInfo: WeaponSlotsInfo {
            mass = LAUNCHER_MASS;
        };
    };

    class BWA3_PzF3_Loaded: BWA3_PzF3 {
        author = "$STR_BWA3_Author";
        scope = 2;
        scopeArsenal = 2;
        baseWeapon = "BWA3_PzF3_Loaded";

        magazines[] = {"CBA_FakeLauncherMagazine"};

        class WeaponSlotsInfo: WeaponSlotsInfo {
            mass = LAUNCHER_MASS + LAUNCHER_MAGAZINE_MASS;
        };
    };

    class BWA3_PzF3_Used: BWA3_PzF3 {
        author = "$STR_BWA3_Author";
        scope = 1;
        scopeArsenal = 1;
        baseWeapon = "BWA3_PzF3_Used";

        magazines[] = {"CBA_FakeLauncherMagazine"};

        class WeaponSlotsInfo: WeaponSlotsInfo {
            mass = LAUNCHER_MASS;
        };
    };
};
```

</details>

**Comment by commy2** [comment]
Closes: https://github.com/acemod/ACE3/issues/1335

**Comment by commy2** [comment]
Doc: https://github.com/CBATeam/CBA_A3/wiki/Disposable-Launchers

File: addons/optics/script_component.hpp [hunk 1]
```diff
@@ -19,6 +19,32 @@
 
 #define PARSE(value) (call compile format ["%1", value])
 #define AMBIENT_BRIGHTNESS (sunOrMoon * sunOrMoon * (1 - overcast * 0.25) + moonIntensity / 5 * (1 - overcast) min 1) // idea by Falke
+#define WEAPON_MAGAZINES(unit,weapon) (weaponsItems (unit) select {_x select 0 == (weapon)} param [0, []] select {_x isEqualType []})
+
+// addWeapon, but without leeching magazines
+#define ADD_GUN(unit,gun) (unit) setUnitLoadout [\
```
Comment (PabstMirror) [inline addons/optics/script_component.hpp]: `setUnitLoadout` breaks anything that uses unique mag IDs
ref: https://github.com/CBATeam/CBA_A3/blob/master/addons/common/fnc_getMagazineIndex.sqf

Comment (commy2) [reply addons/optics/script_component.hpp]: What alternative do we have? These macros are currently only used by the G36A1 and G36A2 and variants.

**Comment by commy2** [comment]
done

**Comment by PabstMirror** [comment]
These might be ace specific, not sure

put down "Launcher [Nato]" box, take nlaw and it is unloaded and cannot fire
nlaw mags don't show, might be because of 
```
    class NLAW_F: CA_LauncherMagazine {
        scope = 1;
```
2 nalws show in ace arsenal, might want
`scopeArsenal = 0;` instead of 1

File: addons/disposable/fnc_firedDisposable.sqf [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        class EventHandlers {
+            fired = "_this call CBA_fnc_firedDisposable";
+        };
+    (end)
+
+Author:
+    commy2
+---------------------------------------------------------------------------- */
+
+params ["_unit", "_launcher", "_muzzle", "", "", "", "_projectile"];
```
Comment (PabstMirror) [inline addons/disposable/fnc_firedDisposable.sqf]: can be done in a later release, but we should add some way of handling spotting rounds (LAW 80)

Comment (commy2) [reply addons/disposable/fnc_firedDisposable.sqf]: Yeah, this is why I don't like dropping the launcher. It needs `addWeaponWithAttachmentsCargoGlobal` to work.

Comment (PabstMirror) [reply addons/disposable/fnc_firedDisposable.sqf]: maybe
```
_wh = "GroundWeaponHolder_Scripted" createVehicle position _unit;
_unit action ["DropWeapon", _wh, _usedLauncher];
```

Comment (commy2) [reply addons/disposable/fnc_firedDisposable.sqf]: DropWeapon is broken. Always has been always will be.

**Comment by commy2** [comment]
Ref https://github.com/acemod/ACE3/pull/6955

**Comment by Dystopian** [comment]
There was a feature in ACE2 - replace also rockets with launchers in ammo crates. So if there are NLAW launcher and 3 its rockets in a box, they become 4 NLAW disposable tubes. Is it a good idea for this PR, maybe optional with setting?

**Comment by commy2** [comment]
Do we want this to be scripted though or just edit the configs in ACE?

**Comment by Dystopian** [comment]
Config doesn't look universal (and quite big), scripted version could handle all disposable launchers/rockets. So I for script.

**Comment by commy2** [comment]
Made it via script, but see no point in having a setting.

File: addons/optics/XEH_preStart.sqf [hunk 3]
```diff
@@ -1,11 +1,12 @@
...
 
+if (configProperties [configFile >> "CBA_PIPItems"] isEqualTo [] && {
+    configProperties [configFile >> "CBA_CarryHandleTypes"] isEqualTo []
+}) exitWith {};
+
+#include "XEH_PREP.sqf"
+
 if (!hasInterface) exitWith {};
 
 // Do not move this below these checks, or this missing function can be exploited.
```
Comment (Dystopian) [inline addons/optics/XEH_preStart.sqf]: looks like this comment became false with this commit

**Comment by Dystopian** [comment]
Personally I like replacing rockets with launchers. Setting would suit those who doesn't like.

File: addons/optics/XEH_preStart.sqf (hunk 3 above)
Comment (commy2) [reply addons/optics/XEH_preStart.sqf]: Yes, sorry I am not done with this PR yet.

**Comment by commy2** [comment]
- adds `CBA_fnc_addWeaponWithoutMagazines`, which is basically `addWeapon` without taking a magazine from inventory (has to be manually reloaded with `addWeaponItem` if needed).

File: addons/disposable/fnc_firedDisposable.sqf (hunk 2 above)
Comment (dedmen) [reply addons/disposable/fnc_firedDisposable.sqf]: `addWeaponWithAttachmentsCargo` will be fully implemented in 1.96 (contact DLC is 1.94)

**Pull request merged**